LSLib/Granny/ClothUtils.cs

```diff
@@ -0,0 +1,384 @@
+            Debug.WriteLine($"GetTargetClothVertices {stopwatch.Elapsed}");
+
+            if (physicsClothVertices.Length == 0 || targetClothVertices.Length == 0)
+            {
+                return [];
+            }
+
+            var kdTree = BuildKdTree(physicsClothVertices);
+            Debug.WriteLine($"BuildKdTree {stopwatch.Elapsed}");
+
+            var triplets = new Triplet[targetClothVertices.Length];
+
+            Parallel.For(0, targetClothVertices.Length, (index) =>
+            {
+                var vertex = targetClothVertices[index];
+                Span<(short PhysicsIndex, float Distance)> triplet = stackalloc (short, float)[3];
+                var tripletIndex = 0;
+
+                // TODO: Searching the whole tree every time is slow. Maybe do RadialSearch? Hard to tell if there was a limit used by the game devs.
+                foreach (var physicsVertex in kdTree.NearestNeighbors(vertex.PositionAsArray, kdTree.Count).Select(t => t.Item2))
+                {
+                    if (vertex.Mask != 0 && (physicsVertex.Mask & vertex.Mask) == 0)
+                    {
+                        continue;
+                    }
+
+                    var distance = (vertex.Position - physicsVertex.Position).Length;
+
+                    // TODO: this doesn't make a whole lot of sense but gets us very close to what the game does
+                    if (tripletIndex == 1 && distance > triplet[0].Distance * 2.875)
+                    {
+                        break;
+                    }
+                    else if (tripletIndex == 2 && (distance > (triplet[0].Distance + triplet[1].Distance) || distance > triplet[0].Distance * 2.8))
+                    {
+                        break;
+                    }
+
+                    triplet[tripletIndex++] = (physicsVertex.PhysicsIndex, distance);
+
+                    if (tripletIndex == 3)
+                    {
+                        break;
+                    }
+                }
+
+                while (tripletIndex < 3)
+                {
+                    triplet[tripletIndex++] = (-1, -1);
+                }
+
+                triplets[index] = new Triplet(triplet[0].PhysicsIndex, triplet[1].PhysicsIndex, triplet[2].PhysicsIndex);
+            });
+
+            Debug.WriteLine($"Generate End {stopwatch.Elapsed}");
+
+            return triplets;
+        }
+
+        private static KDTree<float, T> BuildKdTree<T>(T[] vertices)
+            where T : BaseVertex
+        {
+            // TODO: This kinda sucks. It'd be nice if the KD tree could use the position property directly. Might be worth modifying the package?
+            var points = new float[vertices.Length][];
+
+            for (int i = 0; i < points.Length; i++)
+            {
+                points[i] = vertices[i].PositionAsArray;
+            }
+
+            // don't use Math.Sqrt and Math.Pow for the metric since this'll be called thousands of times (square distance is fine here)
```

Comment:
`var distanceSquaredToTarget = this.Metric(closestPointInFurtherRect, target);`
the KD lib actually expects the returned metric to be squared distance, so this seems to be the intended use :)